apollo-router/src/plugins/authorization/mod.rs

```diff
@@ -1,31 +1,323 @@
+                            referenced_fields_by_type: Default::default(),
+                        },
+                    }));
+                }
+                let mut compiler = ApolloCompiler::new();
+                compiler.set_type_system_hir(schema.type_system.clone());
+                let _id = compiler.add_executable(&query, "query");
+
+                match Self::scopes_filter_query(&compiler, &scopes)? {
+                    None => Ok(Some((query, paths, Arc::new(Mutex::new(compiler))))),
+                    Some((new_query, new_paths)) => {
+                        let mut compiler = ApolloCompiler::new();
+                        compiler.set_type_system_hir(schema.type_system.clone());
+                        let _id = compiler.add_executable(&new_query, "query");
+                        paths.extend(new_paths.into_iter());
+                        Ok(Some((new_query, paths, Arc::new(Mutex::new(compiler)))))
+                    }
+                }
+            }
+        }?;
+
+        match filter_res {
+            None => Ok(None),
+            Some((filtered_query, paths, _)) => {
+                if filtered_query.is_empty() {
+                    Err(QueryPlannerError::PlanningErrors(PlanErrors {
+                        errors: Arc::new(vec![router_bridge::planner::PlanError {
+                            message: Some("Unauthorized query".to_string()),
+                            extensions: None,
+                            validation_error: false,
+                        }]),
+                        usage_reporting: UsageReporting {
+                            stats_report_key: GRAPHQL_VALIDATION_FAILURE_ERROR_KEY.to_string(),
+                            referenced_fields_by_type: Default::default(),
+                        },
+                    }))
+                } else {
+                    let mut compiler = ApolloCompiler::new();
+                    compiler.set_type_system_hir(schema.type_system.clone());
+                    let _id = compiler.add_executable(&filtered_query, "query");
+                    Ok(Some((
+                        filtered_query,
+                        paths,
+                        Arc::new(Mutex::new(compiler)),
+                    )))
+                }
+            }
+        }
+    }
+
+    fn authenticated_filter_query(
+        compiler: &ApolloCompiler,
+        is_authenticated: bool,
+    ) -> Result<Option<(String, Vec<Path>)>, QueryPlannerError> {
+        let id = compiler
+            .db
+            .executable_definition_files()
+            .pop()
+            .expect("the query was added to the compiler earlier");
+
+        let mut visitor = AuthenticatedVisitor::new(compiler, id);
+        let modified_query = transform::document(&mut visitor, id)
+            .map_err(|e| SpecError::ParsingError(e.to_string()))?
+            .to_string();
+
+        if visitor.query_requires_authentication {
```

Comment:
maybe make this logic easier to understand (we return `Ok(None)` if we don't need to modify the query, which means either the query does not contain fields requiring authentication, or we are authenticated and the query does not need to be modified)